### DISCUSSION TITLE
New features development for IHP OpenPDK

### DISCUSSION THREAD

**Discussion by sergeiandreyev** [ra3xdh/qucs_s#876] [Ideas]
As part of the IHP OpenPDK activities we are going to develop the interface with KLayout to enable such features as LVS, Schematic driven layout, etc.
As a first step we're looking to enable CDL netlist export from Qucs-S.
All these developments, of course, will be open source and we're open to suggestions and contributions as well.
This post has the main goal to start discussion and gather feedback (f.e. if somebody is already working on it, or has some nice ideas)

**Comment by ra3xdh** [comment]
Yes, the CDL export implementation is technically possible. I will accept such extension. You have to implement something like `Component::CDL_netlist` that will generate CDL entry. You may look at `verilogawriter.cpp` how it is implemented for Verilog-A export. The simulations and postprocessor equations need not to be exported in CDL and this will simplify the task. I don't remember if CDL allows parametric netlist or the parameters (`.PARAM`) must be resolved.

**Comment by ra3xdh** [comment]
Also keep in mind that the `Component` class APIs are in the refactoring process. See #748 and #858.

**Comment by hpretl** [comment]
Just dreaming: Having pcells for various mm-wave components like T-lines, hybrids, etc. When entering parameters, an EM simulation can be triggered on an auto-generated layout, and the resulting S-parameters can be directly used for circuit simulation. If this works, it could be coupled with an optimizer algorithm to tune the circuit.

**Reply by hpretl** [reply]
And the existing layouts cam then also be used for layout generation.

**Comment by KrzysztofHerman** [comment]
Considering the request presented in #849 maybe we could get some inspiration from the actually existing flows and PDKs. The good news is that all three are following the same structure.  Would be welcome to have:

1.  An alternative way for symbol definition, which is not related to the SPICE/Verilog/VHDL. In `xschem` it is just geometry definition + parametrizable text to represent prefix, model name, pinouts and parameter=value pairs. Maybe an import from `xschem` would be a way to go.  
2. Support of environmental variables like `$PDK_ROOT` and `$PDK` in blocks such as `.LIB` or `.INCLUDE`, what would enable easy access to the models.
3. Mixed mode simulation using XSPICE primitives such as  `dprocess` and `dcosim` and `verilator`. AFIK the development has already started and it is going to be available in 24.4.0 release.  
4. `Transcalc` module extension or separate calculator tool for analyzing/synthetizing transmission lines at IC level for a specified BEOL.

**Comment by ra3xdh** [comment]
> Maybe an import from xschem would be a way to go.

Adding support of xschem symbols should be technically possible. Such extension would be accepted. Also look at https://github.com/Qucs/qucs/issues/659 It defines XML device instead of hardcoded C++ devices. 

> Support of environmental variables like $PDK_ROOT and $PDK in blocks such as .LIB or .INCLUDE

Does Ngspice support environment variable? Otherwise this should be done at Qucs-S netlister level. It is again technically possible task. 

> Mixed mode simulation using XSPICE primitives such as dprocess and dcosim and verilator

Yes, this is planned as #867. I cannot exactly say when the works on this extension will be started. 

> Transcalc module extension or separate calculator tool for analyzing/synthetizing transmission lines at IC level for

Some extension of transcalc is planned as #354

**Comment by KrzysztofHerman** [comment]
I have some generic idea how a PDK could be imported into the Qucs-S. The possible solution could be a separate entry in the left pane auto-generated using ie. `xml` file. 
![image](https://github.com/user-attachments/assets/39340598-d4a9-42fe-b201-4efef36e5247)

Since the geometries of the symbols are fixed and shared between PDK's it can be sourced from `svg` or `xml` file (now it is hardcoded in C++ code). 
The name and parameters of the device could be also included in a separate `xml`. There are also some settings specific to the tools like output netlist format: `Spice` or `CDL` or location of the simulation models (especially for Xyce since ngspice can be set up using `.spiceinit`).

An ChatGPT generated example is shown below:

```
<PDK name="ihp-sg13g2">
    <DeviceCategory name="lv-nmos">
        <Device name="sg13_lv_nmos" prefix="X" symbol="lv-nmos-4t">
            <Model location="libs.tech/xyce/models/"/>
            <Parameters>
                <Parameter name="w" unit="um" default="0.45">
                    <Description>Width of the NMOS transistor</Description>
                </Parameter>
                <Parameter name="l" unit="um" default="0.13">
                    <Description>Length of the NMOS transistor</Description>
                </Parameter>
                <Parameter name="ng" unit=" " default="1">
                    <Description>Number of gates</Description>
                </Parameter>
                <Parameter name="m" unit=" " default="1">
                    <Description>Multiplier factor</Description>
                </Parameter>
            </Parameters>
        </Device>
    </DeviceCategory>
</PDK>


```
A dialog box which enables the user to edit the parameters is shown below:
![image](https://github.com/user-attachments/assets/9994019f-f78f-493e-93b6-edc3af222e4c)


Since all three PDK's, namely SKY130, GF130 and IHP130 use `PDK_ROOT` and `PDK` environmental variables it would be pretty straightforward to setup the simulation either in `ngspice` or `Xyce`.

**Comment by ra3xdh** [comment]
> The possible solution could be a separate entry in the left pane auto-generated using ie. xml file.

I thinks the https://github.com/Qucs/qucs/issues/659 could be used as the starting point here. It implements adding XML devices in the left panel. It uses Qucs XML-like format, but could be implemented for true XML.

**Comment by ra3xdh** [comment]
See also #1047

**Comment by sergeiandreyev** [comment]
is it possible to implement some kind of callbacks for the device symbols parameters, f.e. to show the `capacitance` or `resistance` values on the schematic based on the capacitor / resistor `w/l` edited by the user?
see https://github.com/IHP-GmbH/IHP-Open-PDK/issues/280

**Comment by KrzysztofHerman** [comment]
@ra3xdh  Since IHP  is targeting two schematic editors, namely `xschem`  and `Qucs-s`  we are pursuing to have similar functionalities in booth, One of it is a feasibility of making calculation based on device parameters. In xschem one use `TCL` functions which are evaluated at runtime.  One of possible solutions would be integration in Qucs-s of some equation parser. Just to give you an example: https://beltoforion.de/en/muparser/index.php#idStart. What do you think about it? @andresmmera is now working on the symbol import from XML import and some fields in the symbol definition could be dedicated for the definition of equations.

**Comment by ra3xdh** [comment]
@KrzysztofHerman , adding dependency on muParser is not a problem. This library presents in all Linux distributions and MSYS2. This way is preferable. I used this library before for another projects and it works fine. If the fixed version of muparser and static build are required, it could be embedded in a source tree.

**Comment by ThomasZecha** [comment]
@ra3xdh Since XML schema validation is dropped from Qt6 we are locking forward for an alternative for the XML-based components. One interesting candiate is the Open Source project https://github.com/codesynthesis-com/xsd. It will comprise schema validation of XML documents (components in our case) as well as mapping of XML documents to generated C++ classes. With this mapping the client code becomes very thin: Loading the XML document and readout the values via the generated classes. It depends on xerces-c (https://xerces.apache.org/xerces-c/).

**Comment by ra3xdh** [comment]
I didn't use the XML schema validation in my projects before and cannot say anything about proposed libraries. Could you explain why the validation is required and Qt XML parser is not sufficient? The best way would be to restrict the Qt tools for XML even if will loss in some functionality. Xerces-C is an XML parser itself. It is not a good solution to have two parsers (Qt XML and XercesC) in the code.

When adding the dependency library it should satisfy the following conditions:

* The library should be available as the binary package in Debian/Fedora/Arch repositories and be presented in the MSYS2. Otherwise we will get a pain when preparing a package, because the compiling library from tarball will be required. It is not straightforward procedure for DEB/RPM packages. 
* The library should be compact. Otherwise the size of Windows package will be significant increased. For example bringing another 50Mb library as dependency is not a good.
* The Python/Node.JS/Perl scripting extensions must be avoided. Otherwise it will be needed to distribute the whole Python with the Windows package.

**Comment by ThomasZecha** [comment]
I think first of all the point why use XSD should be clarified. (The other technical points can be discussed later) 

This new feature 'XML defined components' makes it possible to separate the XML component definition and the maintenance of qucs_s. This means that besides the pre-defined qucs_s XML components a user can define own XML based components and import it to qucs_s from scratch. This requires a clear understanding on the part of both qucs_s and the customer of what such an XML-based component must look like. This is where the major advantages of using xsd come into play:

XML Schema Definition (XSD) is a powerful and standardized way to define the structure, content, and constraints of an XML document. Using XSD has several benefits:

1. Defines a Clear Structure
XSD specifies the allowed elements, attributes, and data types in an XML document. This ensures that the XML adheres to a well-defined format, making it predictable and easier to understand.
2. Data Validation
XSD allows for validation of XML documents against the schema, ensuring the data is correct, consistent, and adheres to the specified rules (e.g., an element must be a string, integer, date, etc.).
It supports constraints like required fields, allowed ranges, and enumeration of valid values.
3. Support for Complex Data Types
XSD supports the definition of complex data types, including nested elements and hierarchical relationships, making it suitable for representing intricate data structures.
4. Namespace Support
XSD supports namespaces, allowing the differentiation between elements with the same name but from different contexts. This is particularly useful in large XML systems.
5. Extensibility
XSD is extensible and allows the creation of reusable components like data types and element groups. These can be shared across multiple schemas to maintain consistency and reduce duplication.
6. Interoperability
Since XSD is a W3C standard, it is widely supported by many tools and technologies, ensuring that XML documents defined using XSD are interoperable across different systems and platforms.
7. Integration with Development Tools
Many Integrated Development Environments (IDEs) and XML tools can generate XML templates, validate XML documents, and even auto-generate code from XSD files, speeding up development.
8. Improved Documentation
XSD serves as a self-describing document that communicates the structure and constraints of XML data to developers and users, effectively serving as documentation.
9. Type Safety
XSD introduces data types, enabling type checking and preventing incorrect or unexpected data (e.g., ensuring that a date field contains valid dates).
10. Error Prevention
By defining strict rules for XML structure and content, XSD helps catch errors early during data creation or transmission, reducing issues in downstream processes.

In summary, XML Schema Definition is essential for defining and enforcing robust data structures in XML, ensuring compatibility, consistency, and ease of use in data exchange and integration.

**Comment by ThomasZecha** [comment]
@ra3xdh  It would be nice if it could be clarified whether further research into XSD support for the XML-based component definition is useful or whether XSD shall be avoided. Thank you.

**Comment by ra3xdh** [comment]
Yes, I understood the advantages of XML schema parser. My main concern is additional dependencies. For example if 3Mb application will have 300Mb dependencies this is not acceptable. It's need to check how the size of AppImage and Windows package will increase. Also it's need to check that the dependencies present in Debian, Fedora, OpenSUSE, Arch, and MSYS2. The manual packaging of dependencies is not acceptable.

**Comment by KrzysztofHerman** [comment]
@ra3xdh Together with @andresmmera and @ThomasZecha we are developing the XML import. Since the vast majority of  the components (includig the geometry) are defined in the C++ code we would like to move it to XML separating the component properties and geometry. This implies some decisions about the location of the XML files for booth: (1) standard geometries and (2) components definitions. In the case of `PDK` or any other custom components and geometries we would like to use an environmental variable to reference its location.  Do you have any comments on that ?
+ @sergeiandreyev

**Reply by ra3xdh** [reply]
Please avoid using the environment variables. The most of Qucs-S users are not familiar with command line tools and will not use PDKs. The setting of environment variables is a pain for Windows. Everything should be configurable from the GUI.

**Comment by KrzysztofHerman** [comment]
@ra3xdh thank you for your opinion.Nevertheless the environmental variable(s) are used in the open source PDK. The three manfacturable PDK's: `SKY130`, `GF180` and `IHP-SG13G2` use `PDK_ROOT` and `PDK` variables to reference all the data locations. Until now I it is not possible to run the complete IC design flow under Windows using open source tools. Also for some reason the proprietary tool also use this approach. Although I agree with you that it should be as simple as possible from the user perspective we somehow have to deal with it. What about the location of the XML files, do you have any suggestion where to store it in the file tree ?

**Reply by ra3xdh** [reply]
The environment variables should not be required for usual installation without PDK. This should be optional. Otherwise the portable version will not work. 

> What about the location of the XML files

The preferred location for XML components is somewhere in `share/qucs-s`. For example `share/qucs-s/components`.

**Comment by andresmmera** [comment]
Just for the record, the XML development is currently being done in this branch [1]. 

I've added the XML libraries to the qucs.qrc file [2], so they're in the qucs-s directory [3]. 

Only some standard components have been added, it'll take some time to populate the libraries and do proper testing.


[1] https://github.com/andresmmera/qucs_s/tree/XML_Components
[2] https://github.com/andresmmera/qucs_s/blob/XML_Components/qucs/qucs.qrc
[3] https://github.com/andresmmera/qucs_s/tree/XML_Components/qucs/Libraries

**Reply by ra3xdh** [reply]
> Just for the record, the XML development is currently being done in this branch [1].

Please open a PR and mark it as WIP. This will allow to see the possible conflicts and test with CI.  

> I've added the XML libraries to the qucs.qrc file 

The embedding XML components defintions in the resources is not acceptable. The XML files should be visible for user.

**Comment by ThomasZecha** [comment]
> Yes, I understood the advantages of XML schema parser. My main concern is additional dependencies. For example if 3Mb application will have 300Mb dependencies this is not acceptable. It's need to check how the size of AppImage and Windows package will increase. Also it's need to check that the dependencies present in Debian, Fedora, OpenSUSE, Arch, and MSYS2. The manual packaging of dependencies is not acceptable.

The preferred XSD solution https://www.codesynthesis.com/products/xsd/ introduces the following dependencies:

1. Header only system needed for build, size 1,7 MByte
2. Shared library _**libxerces-c-3.x**_ from https://xerces.apache.org/xerces-c/ needed for build&runtime, size 3-8MByte installed

(1.) Can be (a) installed separately or (b) integrated into qucs-s
(2.) Binaries are present for a wide range of OS, e.g.:

- Debian: https://packages.debian.org/sid/libxerces-c-dev
- Fedora: https://packages.fedoraproject.org/pkgs/xerces-c/xerces-c/
- OpenSuse: https://software.opensuse.org/package/xerces-c
- Arch: https://archlinux.org/packages/extra/x86_64/xerces-c/
- MSYS2: https://packages.msys2.org/packages/mingw-w64-x86_64-xerces-c
- Windows: https://www.dllme.com/dll/files/xerces-c_3_2

@ra3xdh Any comments? Is that acceptable?

**Comment by ra3xdh** [comment]
Yes, this would be acceptable. I will add xerces-c as the dependency for DEB and RPM packages. The embedding library in the Qucs-S distribution is not required.

**Comment by ThomasZecha** [comment]
@ra3xdh Thank you.
Do I unterstand you right: The header-only system from https://www.codesynthesis.com/products/xsd/ shall not become part of qucs-s sources?

**Comment by ra3xdh** [comment]
The inclusion of XSD sources in Qucs-S repository is not required.

**Comment by ThomasZecha** [comment]
Ok, do you provide them on the github build-system?

**Reply by ra3xdh** [reply]
You should specify it as the build dependency in the deploy.yml. It should be available in Ubuntu repository.

**Comment by ThomasZecha** [comment]
Do you mean this: https://github.com/ra3xdh/qucs_s/tree/current/.github/workflows/deploy.yml?

**Reply by ra3xdh** [reply]
Yes, this file is CI script.

**Comment by ThomasZecha** [comment]
@ra3xdh Currently the deploy-workflow seems to be broken for both windows-runs during '**Set up jobs**' with

> Download action repository 'Minionguyjpro/Inno-Setup-Action@v1.2.5' (SHA:7547f78e92afc13352b946775483c5392b7fda10)
> Error: An action could not be found at the URI 'https://api.github.com/repos/Minionguyjpro/Inno-Setup-Action/zipball/7547f78e92afc13352b946775483c5392b7fda10' (2009:1FC45D:1A8FDAF:1B5E1FD:67ECF3BC)

and

> Download action repository 'Minionguyjpro/Inno-Setup-Action@v1.2.5' (SHA:7547f78e92afc13352b946775483c5392b7fda10)
> Error: An action could not be found at the URI 'https://api.github.com/repos/Minionguyjpro/Inno-Setup-Action/zipball/7547f78e92afc13352b946775483c5392b7fda10' (D00B:379679:1B26BF0:1BF5931:67ECF3C0)

respectively. Any idea whats going wrong?

**Comment by ra3xdh** [comment]
CI works correctly. At least everything was operational in the last week. See the https://github.com/ra3xdh/qucs_s/actions/runs/14147853278/job/39637349628?pr=1266 as the example of its operation. How did you try the CI?

**Comment by ThomasZecha** [comment]
I trigger the actions manually on 'current'-branch, see results: https://github.com/ThomasZecha/qucs_s/actions/runs/14192422765

**Comment by ra3xdh** [comment]
It seems @Minionguyjpro has deleted his Github account and all repositories. Need to check and look for replacement.
